src/ray/raylet/node_manager.cc

```diff
@@ -222,7 +222,8 @@ void NodeManager::KillWorker(std::shared_ptr<Worker> worker) {
   retry_timer->expires_from_now(retry_duration);
   retry_timer->async_wait([retry_timer, worker](const boost::system::error_code &error) {
     RAY_LOG(DEBUG) << "Send SIGKILL to worker, pid=" << worker->Pid();
-    // Force kill worker.
+    // Force kill worker. TODO(rkn): Is there some small danger that the worker
+    // has already died and the PID has been reassigned to a different process?
```

Comment:
would a check make sense here? https://unix.stackexchange.com/questions/181001/how-can-i-kill-a-process-and-be-sure-the-pid-hasnt-been-reused#comment301377_181001

Reply:
I don't think that will help since it seems to kill the process on a timer, whereas we want to only kill it after it has probably already died (just to make sure it definitely died).

Reply:
NodeManager can get DisconnectClient message when a worker dies, so it can record the pid as "already killed", and would not send a kill signal when `KillWorker()` is called?

Reply:
seems like the potential race can only happen when the worker dies shortly after the driver, and when we process the disconnect client message for the driver and tries to kill the worker, the workers dies at this very time and another process starts and takes the worker's pid, before the worker's disconnect client message reaches node manager.

Reply:
Hm, yeah I think something like that might work. When the node manager gets the  DisconnectClient message, do we *know* that the worker has died? Or could it have been some other error?

Reply:
I think it'll be very rare, but I do think it could happen any time we call `kill` on a worker that has already died (which we currently do for all workers that die I believe).

Reply:
Yes.. that's a good point. DisconnectClient message indicates connection error, it's most likely due to worker process is dead, but yes that could possibly be other errors. Though we can possibly use some way to check if the pid still belongs to the worker process when we send the kill signal, there's still a rare chance the wrong process is killed.